qa/L0_http/llm_test.py

```diff
@@ -60,35 +130,199 @@ def test_generate(self):
+        ]
+
+        for inputs, error_msg in invalid_type_inputs:
+            self.generate_expect_failure(self._model_name, inputs, error_msg)
+            self.generate_stream_expect_failure(self._model_name, inputs, error_msg)
 
+    def test_duplicate_inputs(self):
+        dupe_prompt = "input 'PROMPT' already exists in request"
+        dupe_stream = "input 'STREAM' already exists in request"
+        # Use JSON string directly as Python Dict doesn't support duplicate keys
+        invalid_type_inputs = [
+            # One duplicate
+            (
+                '{"PROMPT": "hello", "STREAM": false, "PROMPT": "duplicate"}',
+                dupe_prompt,
+            ),
+            ('{"PROMPT": "hello", "STREAM": false, "STREAM": false}', dupe_stream),
+            # Multiple duplicates, parsed in order
+            (
+                '{"PROMPT": "hello", "STREAM": false, "PROMPT": "duplicate", "STREAM": true}',
+                dupe_prompt,
+            ),
+            (
+                '{"PROMPT": "hello", "STREAM": false, "STREAM": true, "PROMPT": "duplicate"}',
+                dupe_stream,
+            ),
+        ]
+        for inputs, error_msg in invalid_type_inputs:
+            self.generate_expect_failure(self._model_name, inputs, error_msg)
+            self.generate_stream_expect_failure(self._model_name, inputs, error_msg)
+
+    def test_generate_stream_response_error(self):
+        # Setup text-based input
+        text = "hello world"
+        inputs = {"PROMPT": [text], "STREAM": True, "REPETITION": 0, "FAIL_LAST": True}
+        r = self.generate_stream(self._model_name, inputs)
+
+        # With "REPETITION": 0, error will be first response and the HTTP code
+        # will be set properly
+        try:
+            r.raise_for_status()
+        except requests.exceptions.HTTPError as e:
+            self.check_sse_responses(r, [{"error": "An Error Occurred"}])
+
+        # With "REPETITION" > 0, the first response is valid response and set
+        # HTTP code to success, so user must validate each response
+        inputs["REPETITION"] = 1
+        r = self.generate_stream(self._model_name, inputs)
         r.raise_for_status()
 
-        # Validate SSE format
-        self.assertIn("Content-Type", r.headers)
-        self.assertIn("text/event-stream", r.headers["Content-Type"])
+        self.check_sse_responses(r, [{"TEXT": text}, {"error": "An Error Occurred"}])
 
-        # SSE format (data: []) is hard to parse, use helper library for simplicity
-        client = sseclient.SSEClient(r)
-        res_count = 0
-        for i, event in enumerate(client.events()):
-            # Parse event data, join events into a single response
-            data = json.loads(event.data)
-            self.assertIn("TEXT", data)
-            self.assertEqual(text, data["TEXT"])
-            res_count += 1
-        self.assertTrue(rep_count, res_count)
+    def test_race_condition(self):
```

Comment:
Can you comment on the risk of race condition here? Are the requests not independent?